packages/core/src/lib/register.ts

```diff
@@ -0,0 +1,99 @@
+import { PasscodeType } from '@logto/schemas';
+import { Context } from 'koa';
+import { Provider } from 'oidc-provider';
+
+import RequestError from '@/errors/RequestError';
+import { WithUserLogContext } from '@/middleware/koa-user-log';
+import { hasUser, hasUserWithEmail, insertUser } from '@/queries/user';
+import assertThat from '@/utils/assert-that';
+import { emailReg } from '@/utils/regex';
+
+import { createPasscode, sendPasscode, verifyPasscode } from './passcode';
+import { encryptUserPassword, generateUserId } from './user';
+
+const assignRegisterResult = async (ctx: Context, provider: Provider, userId: string) => {
+  const redirectTo = await provider.interactionResult(
+    ctx.req,
+    ctx.res,
+    {
+      login: { accountId: userId },
+    },
+    { mergeWithLastSubmission: false }
+  );
+  ctx.body = { redirectTo };
+};
+
+export const registerWithUsernameAndPassword = async (
+  ctx: WithUserLogContext<Context>,
+  provider: Provider,
+  username: string,
+  password: string
+) => {
+  assertThat(
+    username && password,
+    new RequestError({
+      code: 'session.insufficient_info',
+      status: 400,
+    })
+  );
+  assertThat(
+    !(await hasUser(username)),
+    new RequestError({
+      code: 'user.username_exists',
+      status: 422,
+    })
+  );
+
+  const id = await generateUserId();
+
+  const { passwordEncryptionSalt, passwordEncrypted, passwordEncryptionMethod } =
+    encryptUserPassword(id, password);
+
+  await insertUser({
+    id,
+    username,
+    passwordEncrypted,
+    passwordEncryptionMethod,
+    passwordEncryptionSalt,
+  });
+
+  await assignRegisterResult(ctx, provider, id);
+};
+
+export const sendRegisterWithEmailPasscode = async (ctx: Context, jti: string, email: string) => {
```

Comment:
```suggestion
export const sendPasscodeToEmail = async (ctx: Context, jti: string, email: string) => {
```

since the file itself already named "register"